### PR TITLE
Update the throttle limit from the swagger/openAPI update

### PR DESCRIPTION
**Pull request by VirajSalaka**
related to https://github.com/wso2-enterprise/choreo/issues/18716

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java [hunk 1]
```diff
@@ -261,6 +261,8 @@ public enum ExceptionCodes implements ErrorHandler {
     INVALID_OAS3_FOUND(900762, "Invalid OpenAPI V3 definition found", 400, "Invalid OpenAPI V3 definition found"),
     NO_RESOURCES_FOUND(900763, "No resources found", 404, "API must have at least one resource defined"),
     ERROR_REMOVING_EXAMPLES(900764, "Internal Error While Processing Swagger Definition", 500, "Couldn't remove one or more examples from the swagger definition"),
+    INVALID_THROTTLING_LIMIT_FOUND(900765, "Invalid Throttling Limit Found", 400,
+            "Invalid Throttling Limit Found under x-throttling-limit extension"),
```
Comment (VirajSalaka) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java:265]: single line

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/definitions/OAS2Parser.java [hunk 2]
```diff
@@ -689,6 +689,31 @@ public APIDefinitionValidationResponse validateAPIDefinition(String apiDefinitio
...
+            // Validate the x-throttling-limit extension for operations
+            for (String path : swagger.getPaths().keySet()) {
+                Path pathItem = swagger.getPaths().get(path);
+                for (Operation operation: pathItem.getOperations()) {
+                    if (operation.getVendorExtensions() != null &&
+                            operation.getVendorExtensions().containsKey(xThrottlingLimitExtension)) {
+                        OASParserUtil.validateAPIThrottleLimitExtension(validationResponse,
+                                operation.getVendorExtensions().get(xThrottlingLimitExtension));
+                        if (validationResponse.getErrorItems().size() > 0) {
+                            return validationResponse;
```
Comment (VirajSalaka) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/definitions/OAS2Parser.java:711]: Modify the error item's description with more info like which path, which operation etc.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/definitions/OAS3Parser.java [hunk 3]
```diff
@@ -777,6 +778,32 @@ public APIDefinitionValidationResponse validateAPIDefinition(String apiDefinitio
...
+                    if (operation.getExtensions() != null
+                            && operation.getExtensions().containsKey("x-throttling-limit")) {
+                        OASParserUtil.validateAPIThrottleLimitExtension(validationResponse,
+                                operation.getExtensions().get("x-throttling-limit"));
+                        if (validationResponse.getErrorItems().size() > 0) {
+                            return validationResponse;
+                        }
+                    }
+                }
+                if (validationResponse.getErrorItems().size() > 0) {
```
Comment (VirajSalaka) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/definitions/OAS3Parser.java:802]: we can get rid of this statement.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/definitions/OASParserUtilTest.java [hunk 4]
```diff
@@ -581,4 +582,57 @@ public void testGetOASDefinitionWithTierContentAwareProperty() throws Exception
...
+        validExtension.put("unit", "MINUTE");
+        OASParserUtil.validateAPIThrottleLimitExtension(validationResponse, validExtension);
+        Assert.assertEquals("Having proper extension should not result in an error",
+                0, validationResponse.getErrorItems().size());
+
+        validationResponse = new APIDefinitionValidationResponse();
+        Map<String, Object> missingRequestCount = new HashMap<>();
+        missingRequestCount.put("unit", "Min");
+        OASParserUtil.validateAPIThrottleLimitExtension(validationResponse, missingRequestCount);
+        Assert.assertEquals("Having null extension should not result in an error",
```
Comment (VirajSalaka) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/definitions/OASParserUtilTest.java:604]: Assertion messages needs to be changed.

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java (hunk 1 above)
Comment (VirajSalaka) [reply components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java:265]: nothing to change as it should go to two lines considering the checkstyle guideline we follow

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/definitions/OAS3Parser.java [hunk 5]
```diff
@@ -2049,4 +2081,13 @@ public String getVendorFromExtension(String swaggerContent) {
     public String getType() {
         return null;
     }
+
+    @Override
+    public Object getRootLevelSwaggerExtension(String swaggerContent, String extensionKey) {
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/definitions/OAS3Parser.java:2086]: Removed `swagger` as this method goes into AsyncAPI parser as well
```suggestion
    public Object getRootLevelExtension(String swaggerContent, String extensionKey) {
```

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/definitions/OAS3Parser.java [hunk 6]
```diff
@@ -777,6 +777,38 @@ public APIDefinitionValidationResponse validateAPIDefinition(String apiDefinitio
...
+                for (Map.Entry<PathItem.HttpMethod, Operation> operationEntry :
+                        pathItem.readOperationsMap().entrySet()) {
+                    Operation operation = operationEntry.getValue();
+                    if (operation.getExtensions() != null
+                            && operation.getExtensions().containsKey(xThrottlingLimitExtension)) {
+                        ErrorItem errorItem = OASParserUtil.validateAPIThrottleLimitExtension(
+                                operation.getExtensions().get(xThrottlingLimitExtension));
+                        if (errorItem != null) {
+                            validationResponse.getErrorItems().add(errorItem);
+                            errorItem.setDescription(String.format("Error while validating the x-throttling-limit " +
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/definitions/OAS3Parser.java:804]: This is basically we found a user issue right? Its not really an error during validation rather we located some issues after doing the validation
```suggestion
                            errorItem.setDescription(String.format("Invalid x-throttling-limit " +
```

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIDefinition.java [hunk 7]
```diff
@@ -266,4 +268,40 @@ public abstract String processDisableSecurityExtension(String swaggerContent)
...
+     * @param swaggerContent String
+     * @return {@link ThrottlingLimit}
+     * @throws APIManagementException if the value under extension cannot be mapped to a {@link ThrottlingLimit}
+     *                                object
+     */
+    public ThrottlingLimit getAPILevelThrottlingLimit(String swaggerContent)
+            throws APIManagementException {
+        Object extension = getRootLevelSwaggerExtension(swaggerContent,
+                APIConstants.X_THROTTLING_LIMIT_OPENAPI_EXTENSION);
+        ObjectMapper mapper = new ObjectMapper();
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIDefinition.java:294]: Lets replace `swaggerContent` with apiDefinitionContent

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIDefinition.java [hunk 8]
```diff
@@ -266,4 +268,40 @@ public abstract String processDisableSecurityExtension(String swaggerContent)
...
+        try {
+            if (extension != null) {
+                return mapper.convertValue(extension, ThrottlingLimit.class);
+            }
+            // Unreachable condition as the extension is already validated before calling this method.
+        } catch (IllegalArgumentException e) {
+            throw new APIManagementException(
+                    String.format("Invalid content under the %s extension in swagger definition",
+                            APIConstants.X_THROTTLING_LIMIT_OPENAPI_EXTENSION), e);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIDefinition.java:303]: Better not to stick to OpenAPI or Swagger here as this is getting added to AsyncAPI Parser too

**Review by malinthaprasan — APPROVED** [review]

**Review by malinthaprasan — APPROVED** [review]

**Comment by VirajSalaka** [comment]
https://github.com/wso2/carbon-apimgt/pull/11934

Had to revert this PR, because when the API setting are updated from the choreo console, first it calls update-api resource and thereafter it calls the update-swagger resource. Since the swagger definition does not contain all the extensions, the API level rate limit gets removed. We need to update the swagger also with correct extensions and make it identical to the api-properties payload (sent for update-api resource) in order to solve this issue.

**Pull request merged**